packages/fx-core/src/component/error/deployError.ts

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  SystemError,
+  SystemErrorOptions,
+  UserError,
+  UserErrorOptions,
+} from "@microsoft/teamsfx-api";
+import { getDefaultString, getLocalizedString } from "../../common/localizeUtils";
+
+/**
+ * deploy error
+ *
+ * @property name name of error, for classification
+ * @property message always save error message in english
+ * @property errorType which type of error will throw to fx-core
+ * @property displayMessage the message show to user, always save as localized message
+ * @property helpLink link to help page, mostly no use
+ */
+export class BaseDeployInnerError extends Error {
+  errorType: "UserError" | "SystemError";
+  helpLink?: string;
+  displayMessage?: string;
+
+  constructor(
+    errorType: "UserError" | "SystemError",
+    name: string,
+    messageKey?: string,
+    messageParams?: string[],
+    helpLink?: string
+  ) {
+    super(
+      messageKey
+        ? messageParams
+          ? getDefaultString(messageKey, ...messageParams)
+          : getDefaultString(messageKey)
+        : undefined
+    );
+    this.errorType = errorType;
+    this.name = name;
+    this.helpLink = helpLink;
+    this.displayMessage = messageKey
+      ? messageParams
+        ? getLocalizedString(messageKey, ...messageParams)
+        : getLocalizedString(messageKey)
+      : undefined;
+  }
+
+  toFxError(): UserError {
```

Comment:
Since the return type will be either UserError or SystemError, I guess you will update it later